### PR TITLE
VIM-6207: Inject Different Documents Folder's URL Into Export Operation Chain and Descriptor System

### PR DESCRIPTION
**Pull request by supperspidey**
#### Ticket

[VIM-6207](https://vimean.atlassian.net/browse/VIM-6207)

#### Pull Request Checklist

- [x] Resolved any merge conflicts
- [x] No build errors or warnings are introduced
- [x] New files are written in Swift
- [x] New classes contain license headers
- [x] New classes have Documentation
- [x] New public methods have Documentation

#### Issue Summary

This PR enables the descriptor system and the export operation chain to be able to find archives/video files in different Documents folder's URL.

#### Implementation Summary

For all classes that persist descriptors to disk, they are now able to save/load descriptors from a different Documents folder's URL.

`ExportOperation` is now able to save the exported video file to a different Documents folder's URL; however, the injected URL argument into the export operation chain is default to `nil` (a.k.a app's sandbox) and should be kept that way for third-party apps. The reason is because we don't actually save upload descriptors' URLs to disk. Since the SHAs of both the shared container and the app's sandbox change every time the client launches, doing otherwise will render those URLs obsolete. Therefore, we always refer to the main app's Documents folder as the safest place to retrieve video files.

#### Reviewer Tips

N/A

#### How to Test

- Run both sample apps. Check that upload still works.

### CODE REVIEW

**Comment by cameo-github** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          6 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/96fad7b3ef432bd381b423671c9b1dccae17462d/VimeoUpload/Upload/Descriptor System/Old Upload/OldUploadDescriptor.swift#L50">VimeoUpload/Upload/Descriptor System/Old Upload/OldUploadDescriptor.swift#L50</a></strong>: string interpolation produces a debug description for an optional value; did you mean to make this explicit?  <br /><code>
            print("\(self.currentRequest.rawValue) \(self.identifier)")
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/96fad7b3ef432bd381b423671c9b1dccae17462d/VimeoUpload/UIKit/DescriptorKVObserver.swift#L31">VimeoUpload/UIKit/DescriptorKVObserver.swift#L31</a></strong>: extraneous duplicate parameter name; ‘descriptorType’ already has an argument label  <br /><code>
    func descriptorStateDidChange(descriptorType descriptorType: VideoDescriptorType, descriptorState: DescriptorState, isCancelled: Bool, descriptorError: NSError?)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/96fad7b3ef432bd381b423671c9b1dccae17462d/VimeoUpload/UIKit/DescriptorKVObserver.swift#L32">VimeoUpload/UIKit/DescriptorKVObserver.swift#L32</a></strong>: extraneous duplicate parameter name; ‘descriptorType’ already has an argument label  <br /><code>
    func descriptorProgressDidChange(descriptorType descriptorType: VideoDescriptorType, progress: Double)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/96fad7b3ef432bd381b423671c9b1dccae17462d/VimeoUpload/Descriptor System/DescriptorManager.swift#L232">VimeoUpload/Descriptor System/DescriptorManager.swift#L232</a></strong>: conditional downcast from ‘Error?’ to ‘NSError’ is a bridging conversion; did you mean to use ‘as’?  <br /><code>
                let isConnectionError = ((task.error as? NSError)?.isConnectionError() == true || (error as? NSError)?.isConnectionError() == true)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/96fad7b3ef432bd381b423671c9b1dccae17462d/VimeoUpload/Upload/Extensions/AVAsset+Extensions.swift#L32">VimeoUpload/Upload/Extensions/AVAsset+Extensions.swift#L32</a></strong>: extraneous duplicate parameter name; ‘completion’ already has an argument label  <br /><code>
    func approximateFileSize(completion completion: @escaping FloatBlock)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><strong><a href="https://github.com/vimeo/VimeoUpload/blob/96fad7b3ef432bd381b423671c9b1dccae17462d/VimeoUpload/Upload/Descriptor System/Cameo/CAMUploadDescriptor.swift#L27">VimeoUpload/Upload/Descriptor System/Cameo/CAMUploadDescriptor.swift#L27</a></strong>: string interpolation produces a debug description for an optional value; did you mean to make this explicit?  <br /><code>
            print("\(self.currentRequest.rawValue) \(self.identifier)")
</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.845 (0.847) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.350 (0.353) seconds</td>
    </tr>
  </tbody>
</table>

See build details on [CircleCI](https://circleci.com/gh/vimeo/VimeoUpload/290)

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: VimeoUpload/Descriptor System/DescriptorManagerArchiver.swift [hunk 1]
```diff
@@ -47,49 +47,53 @@ class DescriptorManagerArchiver
 
     // MARK: - Initialization
     
-    init(name: String)
+    init(name: String, documentsFolderURL: URL)
```
Comment (jenniferlim) [inline VimeoUpload/Descriptor System/DescriptorManagerArchiver.swift]: Can we make the initializer failable? It looks that the `archiver` is mandatory for this class.

File: VimeoUpload/Descriptor System/DescriptorManagerArchiver.swift [hunk 2]
```diff
@@ -35,7 +35,7 @@ class DescriptorManagerArchiver
 
     // MARK: 
     
-    private let archiver: KeyedArchiver
+    private let archiver: KeyedArchiver?
```
Comment (jenniferlim) [inline VimeoUpload/Descriptor System/DescriptorManagerArchiver.swift]: Does it make sense to instantiate the `DescriptorManagerArchiver` without the `archiver`? It looks like we want this guy non optional. We prolly want the initializer being failable?

File: VimeoUpload/Descriptor System/ReachableDescriptorManager.swift [hunk 3]
```diff
@@ -47,7 +47,24 @@ import VimeoNetworking
...
+    ///   - backgroundSessionIdentifier: An ID of the background upload
+    ///   session.
+    ///   - sharedContainerIdentifier: An ID of a shared sandbox. By default
+    ///   this value is `nil`, but if `VimeoUpload` is used in an app
+    ///   extension, this value must be set.
+    ///   - descriptorManagerDelegate: A delegate object of this descriptor
+    ///   manager.
+    ///   - accessTokenProvider: A closure that provides an authenticated
+    ///   token. Any upload needs this token in order to work properly.
+    ///   - apiVersion: The API version to use.
```
Comment (jenniferlim) [inline VimeoUpload/Descriptor System/ReachableDescriptorManager.swift]: 🌟

File: VimeoUpload/VimeoUploader.swift [hunk 4]
```diff
@@ -56,10 +56,17 @@ open class VimeoUploader<T: VideoDescriptor>
...
+            let documentsFolderURL = try FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: false)
+            self.deletionManager = VideoDeletionManager(sessionManager: self.foregroundSessionManager, documentsFolderURL: documentsFolderURL)
         
-        self.descriptorManager = ReachableDescriptorManager(name: type(of: self).Name, backgroundSessionIdentifier: backgroundSessionIdentifier, descriptorManagerDelegate: descriptorManagerDelegate,
-                                                            accessTokenProvider: accessTokenProvider, apiVersion: apiVersion)
+            self.descriptorManager = ReachableDescriptorManager(name: type(of: self).Name, documentsFolderURL: documentsFolderURL, backgroundSessionIdentifier: backgroundSessionIdentifier, descriptorManagerDelegate: descriptorManagerDelegate, accessTokenProvider: accessTokenProvider, apiVersion: apiVersion)
+        }
+        catch
+        {
+            fatalError("Failure: Documents folder not found")
```
Comment (jenniferlim) [inline VimeoUpload/VimeoUploader.swift]: How about asserting?

**Comment by supperspidey** [comment]
Merci @jenniferlim !

**Pull request merged**